### DISCUSSION TITLE
Nested Rules

### DISCUSSION THREAD

**Discussion by aliirim** [microsoft/RulesEngine#559] [Q&A]
I want to keep the workflow rule results in a rule in the script (no need for business details). For this reason, I created an example script as follows. There are examples of nested rules in the documentation, but they are generally only 1 level.
Does the rule engine allow 2 levels of nested rule as below?
Does using it this way cause performance problems, incorrect results, etc.?

<!DOCTYPE html>
<html lang="en">

<body>
    <pre>
{
    "WorkflowName": "Discount",
    "Rules": [
        {
            "RuleName": "DiscountResult",
            "Rules": [
                {
                    "RuleName": "R1AndR2AndR3",
                    "Rules": [
                        {
                            "RuleName": "R1",
                            "ErrorMessage": "One or more adjust rules failed.",
                            "ErrorType": "Error",
                            "RuleExpressionType": "LambdaExpression",
                            "Expression": "input1.loyaltyFactor = 4"
                        },
                        {
                            "RuleName": "R2",
                            "ErrorMessage": "One or more adjust rules failed.",
                            "ErrorType": "Error",
                            "RuleExpressionType": "LambdaExpression",
                            "Expression": "input1.totalPurchasesToDate >= 100000"
                        },
                        {
                            "RuleName": "R3",
                            "ErrorMessage": "One or more adjust rules failed.",
                            "ErrorType": "Error",
                            "RuleExpressionType": "LambdaExpression",
                            "Expression": "input2.totalOrders > 15 AND input3.noOfVisitsPerMonth > 25"
                        }
                    ],
                    "SuccessEvent": "35",
                    "ErrorMessage": "One or more adjust rules failed.",
                    "ErrorType": "Error",
                    "Operator": "AndAlso"
                },
                {
                    "RuleName": "R4OrR5",
                    "Rules": [
                        {
                            "RuleName": "R4",
                            "ErrorMessage": "One or more adjust rules failed.",
                            "ErrorType": "Error",
                            "RuleExpressionType": "LambdaExpression",
                            "Expression": "input1.loyaltyFactor > 3 AND input1.totalPurchasesToDate >= 50000 AND input1.totalPurchasesToDate <= 100000"
                        },
                        {
                            "RuleName": "R5",
                            "ErrorMessage": "One or more adjust rules failed.",
                            "ErrorType": "Error",
                            "RuleExpressionType": "LambdaExpression",
                            "Expression": "input2.totalOrders > 15"
                        }
                    ],
                    "SuccessEvent": "30",
                    "ErrorMessage": "One or more adjust rules failed.",
                    "ErrorType": "Error",
                    "Operator": "Or"
                }
            ],
            "SuccessEvent": "30",
            "ErrorMessage": "One or more adjust rules failed.",
            "ErrorType": "Error",
            "Operator": "AndAlso"
        }
    ]
}
    </pre>
</body>

</html>

**Comment by ghost** [comment]
Yes I also need help on this. It seems that "Expression" of a rule is ignored when there is any child rule in "Rules" with "Operator". Then it becomes tricky that how you handle 2 or more levels of rules with mixed and/or conditions
